Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -583,9 +583,75 @@ void GraphicsContextSkia::drawLinesForText(const FloatPoint& point, float thickn
         fillRect(dash, localStrokeColor);
 }
 
-void GraphicsContextSkia::drawDotsForDocumentMarker(const FloatRect&, DocumentMarkerLineStyle)
+void GraphicsContextSkia::drawDotsForDocumentMarker(const FloatRect& boundaries, DocumentMarkerLineStyle style)
 {
-    notImplemented();
+    if (style.mode != DocumentMarkerLineStyleMode::Spelling
+        && style.mode != DocumentMarkerLineStyleMode::Grammar)
+        return;
+
+    canvas().drawPath(createFillableTriangleWave(boundaries), createFillPaint(static_cast<SkColor>(style.color)));
+}
+
+// Creates a path comprising of two triangle waves separated by some empty space in Y axis.
+// The empty space can be filled using SkPaint::kFill_Style thus forming an elegant triangle wave.
+// Such triangle wave can be used e.g. as an error underline for text.
+SkPath GraphicsContextSkia::createFillableTriangleWave(const FloatRect& boundaries)
```

Comment:
This doesn't need to be a static member of GraphicsContextSkia, it could be just a static function or a lambda inside `drawDotsForDocumentMarker` since it's only needed there. I would call it `createErrorUnderlinePath`